File: app/src/main/java/com/x7/steambadger/LoaderTask.java
package com.x7.steambadger;

import android.app.ProgressDialog;
import android.content.Context;
import android.content.DialogInterface;
import android.os.AsyncTask;

public abstract class LoaderTask<C extends Context> extends AsyncTask<C, Object, Object> {
    
    protected C context;
    protected ProgressDialog dialog;
    
    public LoaderTask(C context) {
        this(context, true);
    }

    public LoaderTask(C context, boolean dialog) {
        this.context = context;
        
        if (dialog) {
        	this.dialog = new ProgressDialog(context);
        	this.dialog.setMessage("Carregando...");
        	this.dialog.setCancelable(true);
        	this.dialog.setOnCancelListener(new DialogInterface.OnCancelListener() {
				
				public void onCancel(DialogInterface dialog) {
					onCancelTriggered();
				}
			});
        	this.dialog.show();
        }
        
        execute(this.context);
    }
    
    public C getContext() {
        return context;
    }

    @Override
    protected final void onProgressUpdate(Object... values) {
    	onUpdate(values);
    }
    
    public void onUpdate(Object... values) {
    }
    
    public final void doUpdate(Object... values) {
    	publishProgress(values);
    }
    
    public final void setCancellable(boolean cancellable) {
    	if (this.dialog != null) {
    		this.dialog.setCancelable(cancellable);
    	}
    }

    @SafeVarargs
    public final C doInBackground(C... params) {
        process();
        return null;
    }
    
    public abstract void process();

    @Override
    protected final void onPostExecute(Object result) {
        onComplete();
        if (dialog != null) {
        	dialog.dismiss();
    	}
    }
    
    public void onComplete() {
    }
    
    public void onCancelTriggered() {
    }
}

File: app/src/main/java/com/dougmaitelli/steambadger/util/LoaderTask.kt
package com.dougmaitelli.steambadger.util

import android.app.Activity
import android.os.AsyncTask
import androidx.appcompat.app.AlertDialog
import com.dougmaitelli.steambadger.R

abstract class LoaderTask<C : Activity> constructor(protected val context: C, showProgress: Boolean = true) : AsyncTask<C, Any, Any>() {

    private var dialog: AlertDialog? = null

    init {
        if (showProgress) {
            val builder = AlertDialog.Builder(context)
            builder.setView(R.layout.progress)
            dialog = builder.create()
            dialog!!.show()
        }

        execute(this.context)
    }

    override fun onProgressUpdate(vararg values: Any) {
        onUpdate(*values)
    }

    open fun onUpdate(vararg values: Any) {}

    fun doUpdate(vararg values: Any) {
        publishProgress(values)
    }

    override fun doInBackground(vararg params: C): C? {
        process()
        return null
    }

    abstract fun process()

    override fun onPostExecute(result: Any?) {
        onComplete()
        if (dialog != null) {
            dialog!!.dismiss()
        }
    }

    open fun onComplete() {}

    open fun onCancelTriggered() {}
}